pkg/validation/seed/validator.go

```diff
@@ -33,71 +33,74 @@ import (
 // ValidateFunc validates a Seed resource
 // On DELETE, the only set fields of seed are Name and Namespace as
 // admissionReview.Request.Object is unset
-type ValidateFunc func(ctx context.Context, seed *kubermaticv1.Seed, op admissionv1beta1.Operation) error
+type validateFunc func(ctx context.Context, seed *kubermaticv1.Seed, op admissionv1beta1.Operation) error
 
-func NewDefaultSeedValidator(
+func newSeedValidator(
 	workerName string,
-	seedsGetter provider.SeedsGetter,
+	client ctrlruntimeclient.Client,
 	seedClientGetter provider.SeedClientGetter,
-) (*Validator, error) {
+) (*validator, error) {
 	labelSelector, err := workerlabel.LabelSelector(workerName)
 	if err != nil {
 		return nil, err
 	}
 	listOpts := &ctrlruntimeclient.ListOptions{LabelSelector: labelSelector}
-	return &Validator{
-		seedsGetter:      seedsGetter,
+	return &validator{
+		client:           client,
 		seedClientGetter: seedClientGetter,
 		lock:             &sync.Mutex{},
 		listOpts:         listOpts,
 	}, nil
 }
 
 // Ensure that Validator.Validate implements ValidateFunc
-var _ ValidateFunc = (&Validator{}).Validate
+var _ validateFunc = (&validator{}).Validate
 
-type Validator struct {
-	seedsGetter      provider.SeedsGetter
+type validator struct {
+	client           ctrlruntimeclient.Client
 	seedClientGetter provider.SeedClientGetter
 	lock             *sync.Mutex
 	// Can be used to insert a labelSelector
 	listOpts *ctrlruntimeclient.ListOptions
 }
 
 // Validate returns an error if the given seed does not pass all validation steps.
-func (sv *Validator) Validate(ctx context.Context, seed *kubermaticv1.Seed, op admissionv1beta1.Operation) error {
+func (v *validator) Validate(ctx context.Context, seed *kubermaticv1.Seed, op admissionv1beta1.Operation) error {
 	// We need locking to make the validation concurrency-safe
 	// (irozzo): this is acceptable as request rate is low, but is it
 	// required?
-	sv.lock.Lock()
-	defer sv.lock.Unlock()
+	v.lock.Lock()
+	defer v.lock.Unlock()
 
-	// (irozzo): Double check why we do need this getter and we cannot rely on
-	// the controller-runtime client only.
-	seeds, err := sv.seedsGetter()
+	seeds := kubermaticv1.SeedList{}
+	err := v.client.List(ctx, &seeds)
 	if err != nil {
 		return fmt.Errorf("failed to get seeds: %v", err)
 	}
+	seedsMap := map[string]*kubermaticv1.Seed{}
+	for _, s := range seeds.Items {
+		seedsMap[s.Name] = &s
+	}
```

Comment:
Are we sure that this works? Isn't `s` re-used in each iteration and so every element in `seedsMap` would en up with the same address?

Reply:
true, this is definitely a bug I'm always assigning the address of s. Good catch.